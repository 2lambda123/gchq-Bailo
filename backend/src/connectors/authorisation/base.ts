import { AccessRequestDoc } from '../../models/AccessRequest.js'
import { FileInterfaceDoc } from '../../models/File.js'
import { EntryVisibility, ModelDoc } from '../../models/Model.js'
import { ReleaseDoc } from '../../models/Release.js'
import { SchemaDoc } from '../../models/Schema.js'
import { UserInterface } from '../../models/User.js'
import { Access, Action } from '../../routes/v1/registryAuth.js'
import { getModelAccessRequestsForUser } from '../../services/accessRequest.js'
import { checkAccessRequestsApproved } from '../../services/review.js'
import { Roles } from '../authentication/Base.js'
import authentication from '../authentication/index.js'
import {
  AccessRequestAction,
  AccessRequestActionKeys,
  FileAction,
  FileActionKeys,
  ImageAction,
  ModelAction,
  ModelActionKeys,
  ReleaseAction,
  ReleaseActionKeys,
  SchemaAction,
  SchemaActionKeys,
} from './actions.js'

type Response = { id: string; success: true } | { id: string; success: false; info: string }

export class BasicAuthorisationConnector {
  async hasModelVisibilityAccess(user: UserInterface, model: ModelDoc) {
    if (model.visibility === EntryVisibility.Public) {
      return true
    }

    const roles = await authentication.getUserModelRoles(user, model)
    if (roles.length === 0) return false

    return true
  }

  async hasApprovedAccessRequest(user: UserInterface, model: ModelDoc) {
    const accessRequests = await getModelAccessRequestsForUser(user, model.id)
    if (accessRequests.length === 0) {
      return false
    }
    return await checkAccessRequestsApproved(accessRequests.map((accessRequest) => accessRequest.id))
  }

  async model(user: UserInterface, model: ModelDoc, action: ModelActionKeys) {
    return (await this.models(user, [model], action))[0]
  }

  async schema(user: UserInterface, schema: SchemaDoc, action: SchemaActionKeys) {
    return (await this.schemas(user, [schema], action))[0]
  }

  async release(user: UserInterface, model: ModelDoc, release: ReleaseDoc, action: ReleaseActionKeys) {
    return (await this.releases(user, model, [release], action))[0]
  }

  async accessRequest(
    user: UserInterface,
    model: ModelDoc,
    accessRequest: AccessRequestDoc,
    action: AccessRequestActionKeys,
  ) {
    return (await this.accessRequests(user, model, [accessRequest], action))[0]
  }

  async file(user: UserInterface, model: ModelDoc, file: FileInterfaceDoc, action: FileActionKeys) {
    return (await this.files(user, model, [file], action))[0]
  }

  async image(user: UserInterface, model: ModelDoc, access: Access) {
    return (await this.images(user, model, [access]))[0]
  }

  async models(user: UserInterface, models: Array<ModelDoc>, action: ModelActionKeys): Promise<Array<Response>> {
    return Promise.all(
      models.map(async (model) => {
        // Prohibit non-collaborators from seeing private models
        if (!(await this.hasModelVisibilityAccess(user, model))) {
          return {
            id: model.id,
            success: false,
            info: 'You cannot interact with a private model that you do not have access to.',
          }
        }

        // Check a user has a role before allowing write actions
        if (
          [ModelAction.Write, ModelAction.Update].some((a) => a === action) &&
          (await missingRequiredRole(user, model, ['owner', 'collaborator']))
        ) {
          return { id: model.id, success: false, info: 'Only owners and collaborators can update a model.' }
        }

        return { id: model.id, success: true }
      }),
    )
  }

  async schemas(user: UserInterface, schemas: Array<SchemaDoc>, action: SchemaActionKeys): Promise<Array<Response>> {
    if (action === SchemaAction.Create || action === SchemaAction.Delete) {
      const isAdmin = await authentication.hasRole(user, Roles.Admin)

      if (!isAdmin) {
        return schemas.map((schema) => ({
          id: schema.id,
          success: false,
          info: 'You cannot upload or modify a schema if you are not an admin.',
        }))
      }
    }

    return schemas.map((schema) => ({
      id: schema.id,
      success: true,
    }))
  }

  async releases(
    user: UserInterface,
    model: ModelDoc,
    releases: Array<ReleaseDoc>,
    action: ReleaseActionKeys,
  ): Promise<Array<Response>> {
    // We don't have any specific roles dedicated to releases, so we pass it through to the model authorisation checker.
    // We do need to map some actions to other model actions.
    const actionMap: Record<ReleaseActionKeys, ModelActionKeys> = {
      [ReleaseAction.Create]: ModelAction.Write,
      [ReleaseAction.Delete]: ModelAction.Write,
      [ReleaseAction.Update]: ModelAction.Update,
      [ReleaseAction.View]: ModelAction.View,
    }

    return new Array(releases.length).fill(await this.model(user, model, actionMap[action]))
  }

  async accessRequests(
    user: UserInterface,
    model: ModelDoc,
    accessRequests: Array<AccessRequestDoc>,
    action: AccessRequestActionKeys,
  ): Promise<Array<Response>> {
    const entities = await authentication.getEntities(user)

    return Promise.all(
      accessRequests.map(async (request) => {
        // Does any individual in the access request share an entity with our user?
        const isNamed = request.metadata.overview.entities.some((value) => entities.includes(value))

        /**
         * Reject if:
         *  - user is not named on the access request
         *  - user is not the owner of the model
         *  - the user is trying to delete or update an existing AR
         */
        if (
          !isNamed &&
          (await missingRequiredRole(user, model, ['owner'])) &&
          [AccessRequestAction.Delete, AccessRequestAction.Update].includes(action)
        ) {
          return { success: false, info: 'You cannot change an access request you do not own', id: request.id }
        }

        // Otherwise they either own the model, access request or this is a read-only action.
        return { success: true, id: request.id }
      }),
    )
  }

  async files(
    user: UserInterface,
    model: ModelDoc,
    files: Array<FileInterfaceDoc>,
    action: FileActionKeys,
  ): Promise<Array<Response>> {
    // Does the user have a valid access request for this model?
    const hasApprovedAccessRequest = await this.hasApprovedAccessRequest(user, model)

    return Promise.all(
      files.map(async (file) => {
        // If they are not listed on the model, don't let them upload or delete files.
        if (
          [FileAction.Delete, FileAction.Upload].includes(action) &&
          (await missingRequiredRole(user, model, ['owner', 'collaborator']))
        ) {
          return {
            success: false,
            info: 'You need to be an owner or collaborator in order to upload a file',
            id: file.id,
          }
        }

        if (
          [FileAction.Download].includes(action) &&
          !model.settings.ungovernedAccess &&
          !hasApprovedAccessRequest &&
          (await missingRequiredRole(user, model, ['owner', 'collaborator', 'consumer']))
        ) {
          return {
            success: false,
            info: 'You need to have an approved access request or be an owner, collaborator or consumer to download a file.',
            id: file.id,
          }
        }

        return { success: true, id: file.id }
      }),
    )
  }

  async images(user: UserInterface, model: ModelDoc, accesses: Array<Access>): Promise<Array<Response>> {
    // Does the user have a valid access request for this model?
    const hasAccessRequest = await this.hasApprovedAccessRequest(user, model)

    return Promise.all(
      accesses.map(async (access) => {
        // Don't allow anything beyond pushing and pulling actions.
        if (
          !access.actions.every((action) => [ImageAction.Push, ImageAction.Pull, ImageAction.List].includes(action))
        ) {
          return {
            success: false,
            info: 'You are not allowed to complete any actions beyond `push` or `pull` on an image.',
            id: access.name,
          }
        }

        // If they are not listed on the model, don't let them upload or delete images.
        if (
          (await missingRequiredRole(user, model, ['owner', 'collaborator'])) &&
          access.actions.includes(ImageAction.Push as Action)
        ) {
          return {
            success: false,
            info: 'You need to be an owner or collaborator in order to upload an image',
            id: access.name,
          }
        }

        if (
          !hasAccessRequest &&
          (await missingRequiredRole(user, model, ['owner', 'collaborator', 'consumer'])) &&
          access.actions.includes(ImageAction.Pull as Action) &&
          !model.settings.ungovernedAccess
        ) {
          return {
            success: false,
            info: 'You need to have an approved access request or be an owner, collaborator or consumer to download an image.',
            id: access.name,
          }
        }

        return { success: true, id: access.name }
      }),
    )
  }
}

async function missingRequiredRole(user: UserInterface, model: ModelDoc, roles: Array<string>) {
  const modelRoles = await authentication.getUserModelRoles(user, model)
  return !modelRoles.some((value) => roles.includes(value))
}

export async function partials<T>(
  data: Array<T>,
  responses: Array<Response>,
  func: (items: Array<T>) => Array<Response>,
): Promise<Array<Response>> {
  const items = data.filter((_, i) => responses[i].success)
  const newResponses = await Promise.resolve(func(items))

  if (newResponses.length !== items.length) {
    throw new Error('The function did not return a response for every item.')
  }

  let responsesIndex = 0
  const mergedResponses = responses.map((response) => {
    if (response.success) {
      return newResponses[responsesIndex++]
    }

    return response
  })

  return mergedResponses
}
