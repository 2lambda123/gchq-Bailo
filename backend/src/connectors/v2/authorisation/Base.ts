import { AccessRequestDoc } from '../../../models/v2/AccessRequest.js'
import { ModelDoc, ModelVisibility } from '../../../models/v2/Model.js'
import { ReleaseDoc } from '../../../models/v2/Release.js'
import { SchemaDoc } from '../../../models/v2/Schema.js'
import { UserDoc } from '../../../models/v2/User.js'
import authentication from '../authentication/index.js'

export const ModelAction = {
  Create: 'create',
  View: 'view',
  Update: 'update',
  DeleteFile: 'delete_file',
  UploadFile: 'upload_file',
  Write: 'write',
} as const
export type ModelActionKeys = (typeof ModelAction)[keyof typeof ModelAction]

export const ReleaseAction = {
  Create: 'create',
  View: 'view',
  Delete: 'delete',
}
export type ReleaseActionKeys = (typeof ReleaseAction)[keyof typeof ReleaseAction]

export const AccessRequestAction = {
  Create: 'create',
  View: 'view',
  Update: 'update',
  Delete: 'delete',
}
export type AccessRequestActionKeys = (typeof AccessRequestAction)[keyof typeof AccessRequestAction]

export const SchemaAction = {
  Create: 'create',
}
export type SchemaActionKeys = (typeof SchemaAction)[keyof typeof SchemaAction]

export abstract class BaseAuthorisationConnector {
  abstract userModelAction(user: UserDoc, model: ModelDoc, action: ModelActionKeys): Promise<boolean>
  abstract userSchemaAction(user: UserDoc, Schema: SchemaDoc, action: SchemaActionKeys): Promise<boolean>
  abstract userReleaseAction(
    user: UserDoc,
    model: ModelDoc,
    release: ReleaseDoc,
    action: ReleaseActionKeys,
  ): Promise<boolean>
  abstract userAccessRequestAction(
    user: UserDoc,
    model: ModelDoc,
    accessRequest: AccessRequestDoc,
    action: AccessRequestActionKeys,
  ): Promise<boolean>

  async hasModelVisibilityAccess(user: UserDoc, model: ModelDoc) {
    if (model.visibility === ModelVisibility.Public) {
      return true
    }

    const roles = await authentication.getUserModelRoles(user, model)
    if (roles.length === 0) return false

    return true
  }
}
