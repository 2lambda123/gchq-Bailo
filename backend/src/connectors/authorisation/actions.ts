export const ModelAction = {
  Create: 'create',
  View: 'view',
  Update: 'update',
  Write: 'write',
} as const
export type ModelActionKeys = (typeof ModelAction)[keyof typeof ModelAction]

export const ReleaseAction = {
  Create: 'create',
  View: 'view',
  Delete: 'delete',
  Update: 'update',
}
export type ReleaseActionKeys = (typeof ReleaseAction)[keyof typeof ReleaseAction]

export const AccessRequestAction = {
  Create: 'create',
  View: 'view',
  Update: 'update',
  Delete: 'delete',
}
export type AccessRequestActionKeys = (typeof AccessRequestAction)[keyof typeof AccessRequestAction]

export const SchemaAction = {
  Create: 'create',
  Delete: 'delete',
  Update: 'update',
}
export type SchemaActionKeys = (typeof SchemaAction)[keyof typeof SchemaAction]

export const FileAction = {
  Delete: 'delete',
  Upload: 'upload',
  // 'view' refers to the ability to see metadata about the file.  'download' lets the user view the file contents.
  View: 'view',
  Download: 'download',
}
export type FileActionKeys = (typeof FileAction)[keyof typeof FileAction]

export const ImageAction = {
  Pull: 'pull',
  Push: 'push',
  List: 'list',
}

export type ImageActionKeys = (typeof ImageAction)[keyof typeof ImageAction]

export const ModelCardAction = {
  Create: 'create',
  View: 'view',
  Update: 'update',
  Write: 'write',
} as const
export type ModelCardActionKeys = (typeof ModelCardAction)[keyof typeof ModelCardAction]
