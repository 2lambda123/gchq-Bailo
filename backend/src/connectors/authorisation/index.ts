import config from '../../utils/config.js'
import { ConfigurationError } from '../../utils/error.js'
import { BasicAuthorisationConnector } from './base.js'

export const AuthorisationKind = {
  Basic: 'basic',
} as const
export type AuthorisationKindKeys = (typeof AuthorisationKind)[keyof typeof AuthorisationKind]

let authorisationConnector: undefined | BasicAuthorisationConnector = undefined
export function getAuthorisationConnector(cache = true): BasicAuthorisationConnector {
  if (authorisationConnector && cache) {
    return authorisationConnector
  }

  switch (config.connectors.authorisation.kind) {
    case AuthorisationKind.Basic:
      authorisationConnector = new BasicAuthorisationConnector()
      break
    default:
      throw ConfigurationError(`'${config.connectors.authorisation.kind}' is not a valid authorisation kind.`, {
        validKinds: Object.values(AuthorisationKind),
      })
  }

  return authorisationConnector
}

export default getAuthorisationConnector()
