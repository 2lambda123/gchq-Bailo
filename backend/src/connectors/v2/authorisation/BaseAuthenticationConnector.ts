import { ModelDoc } from '../../../models/v2/Model.js'
import { ReleaseDoc } from '../../../models/v2/Release.js'
import { UserDoc } from '../../../models/v2/User.js'

export const ModelAction = {
  Create: 'create',
  View: 'view',
  Update: 'update',
  UploadFile: 'upload_file',
  Write: 'write',
} as const
export type ModelActionKeys = (typeof ModelAction)[keyof typeof ModelAction]

export const ReleaseAction = {
  Create: 'create',
  View: 'view',
  Delete: 'delete',
}
export type ReleaseActionKeys = (typeof ReleaseAction)[keyof typeof ReleaseAction]

export abstract class BaseAuthorisationConnector {
  abstract userModelAction(user: UserDoc, model: ModelDoc, action: ModelActionKeys): Promise<boolean>
  abstract userReleaseAction(
    user: UserDoc,
    model: ModelDoc,
    release: ReleaseDoc,
    action: ReleaseActionKeys,
  ): Promise<boolean>
  abstract getEntities(user: UserDoc): Promise<Array<string>>
  abstract getUserInformation(userEntity: string): Promise<{ email: string }>
  abstract getEntityMembers(entity: string): Promise<Array<string>>

  async getUserInformationList(entity): Promise<Promise<{ email: string }>[]> {
    const entities = await this.getEntityMembers(entity)
    return entities.map((member) => this.getUserInformation(member))
  }
}
