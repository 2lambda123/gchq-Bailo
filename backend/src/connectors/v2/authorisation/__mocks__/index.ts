import { vi } from 'vitest'

const auth = {
  hasModelVisibilityAccess: vi.fn(() => true),
  hasAccessRequest: vi.fn(() => true),

  model: vi.fn(() => ({ success: true })),
  modelBatch: vi.fn(() => [{ success: true }]),

  schema: vi.fn(() => ({ success: true })),
  schemaBatch: vi.fn(() => [{ success: true }]),

  accessRequest: vi.fn(() => ({ success: true })),
  accessRequestBatch: vi.fn(() => [{ success: true }]),

  release: vi.fn(() => ({ success: true })),
  releaseBatch: vi.fn(() => [{ success: true }]),

  file: vi.fn(() => ({ success: true })),
  fileBatch: vi.fn(() => [{ success: true }]),

  image: vi.fn(() => ({ success: true })),
  imageBatch: vi.fn(() => [{ success: true }]),
}

export default auth
