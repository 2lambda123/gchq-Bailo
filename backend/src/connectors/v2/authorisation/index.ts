import config from '../../../utils/v2/config.js'
import { BaseAuthorisationConnector } from './Base.js'
import { SillyAuthorisationConnector } from './silly.js'

let authConnector: undefined | BaseAuthorisationConnector = undefined
export function getAuthorisationConnector(cache = true) {
  if (authConnector && cache) {
    return authConnector
  }

  switch (config.connectors.authorisation.kind) {
    case 'silly':
      authConnector = new SillyAuthorisationConnector()
      break
    default:
      throw new Error('No valid authorisation connector provided.')
  }

  return authConnector
}

export default getAuthorisationConnector()
