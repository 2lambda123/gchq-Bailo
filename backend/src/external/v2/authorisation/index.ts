import { ModelDoc } from '../../../models/v2/Model.js'
import { UserDoc } from '../../../models/v2/User.js'
import config from '../../../utils/v2/config.js'
import { SillyAuthorisationConnector } from './silly.js'

export const ModelAction = {
  Create: 'create',
  View: 'view',
} as const

export type ModelActionKeys = (typeof ModelAction)[keyof typeof ModelAction]

export abstract class BaseAuthorisationConnector {
  abstract userModelAction(user: UserDoc, model: ModelDoc, action: ModelActionKeys): Promise<boolean>
}

let authConnector: undefined | BaseAuthorisationConnector = undefined
export function getAuthorisationConnector() {
  if (authConnector) {
    return authConnector
  }

  switch (config.connectors.authorisation.kind) {
    case 'silly':
      authConnector = new SillyAuthorisationConnector()
      break
    default:
      throw new Error('No valid authorisation connector provided.')
  }

  return authConnector
}

export default getAuthorisationConnector()
