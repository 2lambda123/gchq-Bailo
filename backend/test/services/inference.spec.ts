import { describe, expect, test, vi } from 'vitest'

import authorisation from '../../src/connectors/v2/authorisation/index.js'
import {
  createInference,
  getInferenceByImage,
  getInferencesByModel,
  updateInference,
} from '../../src/services/v2/inference.js'

vi.mock('../../src/connectors/v2/authorisation/index.js')

const modelMocks = vi.hoisted(() => ({
  getModelById: vi.fn(),
  getModelCardRevision: vi.fn(),
}))
vi.mock('../../src/services/v2/model.js', () => modelMocks)

const registryMocks = vi.hoisted(() => ({
  listModelImages: vi.fn(() => [{ repository: 'test', name: 'nginx', tags: ['latest'] }]),
}))
vi.mock('../../src/services/v2/registry.js', () => registryMocks)

const inference = {
  image: 'nginx',
  tag: 'latest',
  description: 'test',
  settings: {
    processorType: 'cpu',
    port: 8000,
    memory: 1,
  },
}

const inferenceModelMocks = vi.hoisted(() => {
  const obj: any = {}

  obj.aggregate = vi.fn(() => obj)
  obj.match = vi.fn(() => obj)
  obj.sort = vi.fn(() => obj)
  obj.lookup = vi.fn(() => obj)
  obj.append = vi.fn(() => obj)
  obj.find = vi.fn(() => obj)
  obj.findOne = vi.fn(() => obj)
  obj.updateOne = vi.fn(() => obj)
  obj.updateMany = vi.fn(() => obj)
  obj.save = vi.fn(() => obj)
  obj.delete = vi.fn(() => obj)
  obj.findOneAndUpdate = vi.fn(() => obj)
  obj.filter = vi.fn(() => obj)

  const model: any = vi.fn((params) => ({ ...obj, ...params }))
  Object.assign(model, obj)

  return model
})
vi.mock('../../src/models/v2/Inference.js', () => ({ default: inferenceModelMocks }))

describe('services > inference', () => {
  test('createInference > simple', async () => {
    const mockUser: any = { dn: 'test' }
    await createInference(mockUser, 'test', inference)

    expect(inferenceModelMocks.save).toBeCalled()
    expect(inferenceModelMocks).toBeCalled()
  })

  test('createInference > non-existent image', async () => {
    const mockUser: any = { dn: 'test' }
    expect(() =>
      createInference(mockUser, 'test', { image: 'non-existent', tag: 'image' } as any),
    ).rejects.toThrowError(/^Image non-existent:image was not found on this model./)
  })
  test('updateInference > success', async () => {
    await updateInference({} as any, 'test', 'nginx', 'latest', {
      description: 'New description',
      settings: {
        port: 8000,
        memory: 4,
        processorType: 'cpu',
      },
    })
    expect(inferenceModelMocks.findOneAndUpdate).toBeCalled()
  })
  test('getInferenceByImage > good', async () => {
    await getInferenceByImage({} as any, 'test', 'nginx', 'latest')
    expect(inferenceModelMocks.findOne).toBeCalled()
  })

  test('getInferenceByImage > no permission', async () => {
    inferenceModelMocks.findOne.mockResolvedValue({ image: 'nginx', tag: 'latest' })

    vi.mocked(authorisation.model).mockResolvedValue({
      info: 'You do not have permission to view this inference.',
      success: false,
      id: '',
    })
    expect(() => getInferenceByImage({} as any, 'test', 'non-existant', 'image')).rejects.toThrowError(
      /^You do not have permission to view this inference./,
    )
  })
  test('getInferenceByImage > no inference', async () => {
    inferenceModelMocks.findOne.mockResolvedValueOnce(undefined)
    expect(() => getInferenceByImage({} as any, 'test', 'non-existent', 'image')).rejects.toThrowError(
      /^The requested inferencing service was not found./,
    )
  })

  test('getInferenceByModel > good', async () => {
    modelMocks.getModelById.mockResolvedValue('modelId')
    inferenceModelMocks.find.mockResolvedValue([
      { image: 'nginx', tag: 'latest' },
      { image: 'yolov4', tag: 'latest' },
    ])

    const inference = await getInferencesByModel({} as any, 'modelId')
    expect(inference).toMatchSnapshot()
  })
})
