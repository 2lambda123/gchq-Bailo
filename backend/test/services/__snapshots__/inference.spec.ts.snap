// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > inference > getInferenceByModel > good 1`] = `
[
  {
    "image": "nginx",
    "tag": "latest",
  },
  {
    "image": "yolov4",
    "tag": "latest",
  },
]
`;

exports[`services > release > getModelReleases > good 1`] = `
[
  {
    "modelId": "modelId",
  },
]
`;

exports[`services > release > getModelReleases > good 2`] = `
[
  {
    "updatedAt": -1,
  },
]
`;

exports[`services > release > getModelReleases > good 3`] = `
[
  {
    "as": "model",
    "foreignField": "id",
    "from": "v2_models",
    "localField": "modelId",
  },
]
`;

exports[`services > release > getModelReleases > good 4`] = `
[
  {
    "$set": {
      "model": {
        "$arrayElemAt": [
          "$model",
          0,
        ],
      },
    },
  },
]
`;
